### PR TITLE
#220 Introduced ClusterRepository for loading and storing MesosCluster

### PR DESCRIPTION
**Pull request by frankscholten**
* Moved command logic inside Command objects
* Main only responsibility is cli args parsing
* Added unit test for main
* Added hashcode and equals for MesosCluster
* Lots of little refactorings
* TODO: Add info unit tests and test state and install commands
* TODO: Test command hooks

### CODE REVIEW

File: minimesos/src/main/java/com/containersol/minimesos/main/Main.java [hunk 1]
```diff
@@ -3,153 +3,125 @@
...
-    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
 
-    private static CommandUp commandUp;
-
-    private static String clusterId;
+    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
 
-    @Parameter(names = {"--help", "-help", "-?", "-h"}, help = true)
+    @Parameter(names = {"--help", "-help", "-?", "-h"}, description = "Show help")
     private static boolean help = false;
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/main/Main.java]: Holly Molly!! It took two hours to discover the `static` word, which was killing all tests !!!

File: minimesos/src/test/java/com/containersol/minimesos/main/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    private CommandDestroy commandDestroy;
+    private CommandState commandState;
+    private CommandInstall commandInstall;
+    private CommandUp commandUp;
+
+    @Before
+    public void before() {
+        commandUp = spy(new CommandUp());
+        reset(commandUp);
+        when(commandUp.execute()).thenReturn(mock(MesosCluster.class));
```
Comment (sadovnikov) [inline minimesos/src/test/java/com/containersol/minimesos/main/MainTest.java]: Although I believe mocks are for introducing unit test for legacy code only, this line is wrong: `commandUp.execute()` is invoked and it's result in passed to `when()`. Had to be replaced by
`doReturn(mock(MesosCluster.class)).when(commandUp).execute();`

File: minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterRepository.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.apache.commons.io.IOUtils;
+import org.apache.log4j.Logger;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class ClusterRepository {
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterRepository.java]: I truly don't want to sound like a priest. However, what if we try to javadoc our code

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 4]
```diff
@@ -40,11 +37,13 @@
  */
 public class MesosCluster extends ExternalResource {
 
-    public static final String MINIMESOS_HOST_DIR_PROPERTY = "minimesos.host.dir";
-    public static final String MINIMESOS_FILE_PROPERTY = "minimesos.cluster";
+    private static Logger CLILOGGER = Logger.getLogger(Main.class);
```
Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Please, no CLILOGGER here. Cluster can return data. Printing it out should be handled outside.
`Main` uses `Cluster`. `Cluster` should not use `Main` - cycle dependency

Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: You're right

File: minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterRepository.java (hunk 3 above)
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/ClusterRepository.java]: Agreed

**Comment by sadovnikov** [comment]
Logging needs to be reviewed !!!

**Comment by sadovnikov** [comment]
@frankscholten could you review this PR?

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 5]
```diff
@@ -385,76 +391,12 @@ public static Container getContainer(String clusterId, String role) {
 
     public static String getContainerIp(String clusterId, String role) {
         Container container = getContainer(clusterId, role);
-        if ( container != null ) {
-            return DockerContainersUtil.getIpAddress( dockerClient, container.getId() );
+        if (container != null) {
+            return DockerContainersUtil.getIpAddress(dockerClient, container.getId());
         }
         return null;
     }
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Is getContainerIp still used? Callers should use AbstractContainer.getIpAddress()

File: minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java [hunk 6]
```diff
@@ -513,22 +446,56 @@ public static void printServiceUrl(String clusterId, String serviceName, Minimes
                     default:
                         uri = "Unknown service type '" + serviceName + "'";
                 }
-                LOGGER.info(uri);
-                return;
+
+                out.println( uri );
+                break;
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/cluster/MesosCluster.java]: Coding style - let's standardize on parentheses without whitespace

File: minimesos/src/main/java/com/containersol/minimesos/main/Command.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.containersol.minimesos.main;
+
+public interface Command {
+
+    // TODO: this does not belong here
+    boolean isExposedHostPorts();
+    // TODO: this does not belong here
+    boolean getStartConsul();
+
```
Comment (frankscholten) [inline minimesos/src/main/java/com/containersol/minimesos/main/Command.java]: Are these fields still needed. What about the Command interface itself?

Comment (sadovnikov) [inline minimesos/src/main/java/com/containersol/minimesos/main/Command.java]: I created #225 for this. These methods seem to be introduced together with already removed hooks

**Pull request merged**